README.md
# Time Utility 0.2.1

Time utility is a series of utility methods regarding date and time for python.

```text
Note: This package requires python 3.6 and higher
```

## Installation
```text
pip3 install time_utility
```

## Import
```python
from time_utility import TimeUtility
```

## Usage
The following are the methods available in TimeUtility

<br><br>
* ###`is_naive(target_datetime)`
Checks if the given datetime object is naive.

Parameters:<br>
1. `target_datetime: datetime.datetime` => The target datetime to be checked

Returns: `bool`

Example:
```python
from time_utility import TimeUtility

is_it_naive = TimeUtility.is_naive(TimeUtility.now())
```





<br><br>
* ###`is_aware(target_datetime)`
Checks if the given datetime object is aware

Example:
```python
from time_utility import TimeUtility

is_it_aware = TimeUtility.is_aware(TimeUtility.now())
```






<br><br>
* ####`make_aware(target_datetime, timezone=pytz.utc)`
Returns an aware datetime object based on the naive target_datetime

Parameters:<br>
1. `target_datetime: datetime.datetime` => The target datetime
2. `timezone` [optional] => The preferred timezone. [default: UTC] 

returns `datetime.datetime`

Example:
```python
from time_utility import TimeUtility
from datetime import datetime

aware = TimeUtility.make_aware(datetime.now())
```






<br><br>
* ####`now(timezone=pytz.utc)`
Returns an aware instance of the current datetime

Parameters:<br>
1. `timezone` [Optional] => The preferred timezone [default: UTC]

returns `datetime.datetime`

Example:
```python
from time_utility import TimeUtility

now = TimeUtility.now()
```





<br><br>
* ####`today(timezone=pytz.utc)`
Returns an aware instance of today's date

Parameters:<br>
1. `timezone` [Optional] => The preferred timezone [default: UTC]

returns `datetime.date`

Example:
```python
from time_utility import TimeUtility

today = TimeUtility.today()
```





<br><br>
* ####`get_date_start(timezone=pytz.utc, year = None, month = None, day = None)`
Returns the date and time of the beginning of the day with the given timezone

Parameters:<br>
1. `timezone` [Optional] => The preferred timezone [default: UTC]
2. `year: int` => The target year. [default: None (will be set to current year)]
3. `month: int` => The target month. [default: None (will be set to current month)]
4. `day: int` => The target day. [default: None (will be set to current day)]

returns `datetime.datetime`

Example:
```python
from time_utility import TimeUtility

day_start = TimeUtility.get_date_start(year=2020, month=5)
```




<br><br>
* ####`get_date_end(timezone=pytz.utc, year = None, month = None, day = None)`
Returns the date and time of the ending of the day with the given timezone

Parameters:<br>
1. `timezone` [Optional] => The preferred timezone [default: UTC]
2. `year: int` => The target year. [default: None (will be set to current year)]
3. `month: int` => The target month. [default: None (will be set to current month)]
4. `day: int` => The target day. [default: None (will be set to current day)]

returns `datetime.datetime`

Example:
```python
from time_utility import TimeUtility

day_start = TimeUtility.get_date_end(year=2020, month=5)
```




<br><br>
* ####`get_month_start(timezone=pytz.utc, year = None, month = None)`
Returns the date and time of the beginning of the month with the given timezone

Parameters:<br>
1. `timezone` [Optional] => The preferred timezone [default: UTC]
2. `year: int` => The target year. [default: None (will be set to current year)]
3. `month: int` => The target month. [default: None (will be set to current month)]

returns `datetime.datetime`

Example:
```python
from time_utility import TimeUtility

day_start = TimeUtility.get_month_start(year=2020, month=5)
```



<br><br>
* ####`get_month_end(timezone=pytz.utc, year = None, month = None)`
Returns the date and time of the ending of the month with the given timezone

Parameters:<br>
1. `timezone` [Optional] => The preferred timezone [default: UTC]
2. `year: int` => The target year. [default: None (will be set to current year)]
3. `month: int` => The target month. [default: None (will be set to current month)]

returns `datetime.datetime`

Example:
```python
from time_utility import TimeUtility

day_start = TimeUtility.get_month_end(year=2020, month=5)
```



<br><br>
* ####`get_year_start(timezone=pytz.utc, year = None)`
Returns the date and time of the beginning of the year with the given timezone

Parameters:<br>
1. `timezone` [Optional] => The preferred timezone [default: UTC]
2. `year: int` => The target year. [default: None (will be set to current year)]

returns `datetime.datetime`

Example:
```python
from time_utility import TimeUtility

day_start = TimeUtility.get_year_start(year=2020)
```



<br><br>
* ####`get_year_end(timezone=pytz.utc, year = None)`
Returns the date and time of the ending of the year with the given timezone

Parameters:<br>
1. `timezone` [Optional] => The preferred timezone [default: UTC]
2. `year: int` => The target year. [default: None (will be set to current year)]

returns `datetime.datetime`

Example:
```python
from time_utility import TimeUtility

day_start = TimeUtility.get_year_end(year=2020)
```





<br><br>
* ####`adjust_offset(original_datetime, offset, local_to_utc)`
Creates and returns a new datetime object using the given datetime object with the given offset in minutes.
The offset should be positive for timezones ahead of UTC (e.g. India) and negative for timezones before UTC (e.g. US) <br>
Note that the offset obtained in Javascript via `new Date().getTimezoneOffset()` should be multiplied by -1.

Parameters:<br>
1. `original_datetime: datetime.datetime` => The original datetime to be adjusted
2. `offset: int` => The offset in minutes
3. `local_to_utc` => Set to True if the original datetime is in local time and set to False if the original datetime is in UTC time

returns `datetime.datetime`

Example:
```python
from time_utility import TimeUtility
from datetime import datetime

adjusted_datetime = TimeUtility.adjust_offset(TimeUtility.make_aware(datetime.now()), 330, True)
```




<br><br>
* ####`get_period(year, month, day, period, offset = 0)`
Gets the start and the end datetime of a selected period

Parameters:<br>
1. `year: int` => The selected year
2. `month: int` => The selected month
3. `day: int` => The selected day
4. `period: str` => The selected period. Your options are: `TimeUtility.DAILY`, `TimeUtility.MONTHLY`, and `TimeUtility.ANNUAL`
5. `offset: int` [optional] => The offset from UTC to adjust the start and the end [default: 0]

returns `(datetime.datetime, datetime.datetime)`

Example:
```python
from time_utility import TimeUtility

start, end = TimeUtility.get_period(year=2020, month=2, day=10, period=TimeUtility.MONTHLY, offset=330)
```




<br><br>
* ####`difference(large_time, small_time, time_span = TimeUtility.SECOND)`
Calculates the difference between two datetime object based on the given time-span

Parameters:<br>
1. `large_time: datetime.datetime` => The bigger datetime object of the two
2. `small_time: datetime.datetime` => The smaller datetime object of the two
3. `time_span: str` [Optional] => The time-span unit of time to calculate the difference. The options are `TimeUtility.MICROSECOND`, `TimeUtility.SECOND`, `TimeUtility.MINUTE`, `TimeUtility.HOUR`, `TimeUtility.DAY`. [default: TimeUtility.SECOND] 

returns `int`

Example:
```python
from time_utility import TimeUtility

difference = TimeUtility.difference(TimeUtility.adjust_offset(TimeUtility.now(), 120, False), TimeUtility.now(), TimeUtility.MINUTE)
```




<br><br>
* ####`is_leap_year(year = None)`
Checks if the given year is a leap year or not

Parameters:<br>
1. `year: int` [Optional] => The target year [default: None (will be set to current year)]

returns `bool`

Example:
```python
from time_utility import TimeUtility

is_leap = TimeUtility.is_leap_year(year=2020)
```


changelog.md
# Change Log

## v0.2.1 (2023-09-05)
- Fixed the types

tests/test_time_utility.py
# Python
import unittest
from datetime import datetime, date
from calendar import monthrange

# Third Party
import pytz

# Time Utility
from time_utility import TimeUtility


class TestTimeUtility(unittest.TestCase):

    def test_is_naive(self):
        self.assertTrue(TimeUtility.is_naive(datetime.now()))

    def test_is_aware(self):
        self.assertTrue(TimeUtility.is_aware(TimeUtility.now()))

    def test_make_aware(self):
        self.assertTrue(TimeUtility.is_aware(TimeUtility.make_aware(datetime.now())))

    def test_now(self):
        self.assertEqual(type(TimeUtility.now()), type(datetime.now()))

    def test_today(self):
        self.assertEqual(type(TimeUtility.today()), type(datetime.now().date()))

    def test_get_date_start(self):
        start = TimeUtility.get_date_start()
        manual = datetime.now(tz=pytz.utc)

        self.assertEqual(start.year, manual.year)
        self.assertEqual(start.month, manual.month)
        self.assertEqual(start.day, manual.day)
        self.assertEqual(start.hour, 0)
        self.assertEqual(start.minute, 0)
        self.assertEqual(start.second, 0)
        self.assertEqual(start.microsecond, 0)

    def test_get_date_end(self):
        start = TimeUtility.get_date_end()
        manual = datetime.now(tz=pytz.utc)

        self.assertEqual(start.year, manual.year)
        self.assertEqual(start.month, manual.month)
        self.assertEqual(start.day, manual.day)
        self.assertEqual(start.hour, 23)
        self.assertEqual(start.minute, 59)
        self.assertEqual(start.second, 59)
        self.assertEqual(start.microsecond, 999999)

    def test_get_month_start(self):
        start = TimeUtility.get_month_start()
        manual = datetime.now(tz=pytz.utc)

        self.assertEqual(start.year, manual.year)
        self.assertEqual(start.month, manual.month)
        self.assertEqual(start.day, 1)
        self.assertEqual(start.hour, 0)
        self.assertEqual(start.minute, 0)
        self.assertEqual(start.second, 0)
        self.assertEqual(start.microsecond, 0)

    def test_get_month_end(self):
        start = TimeUtility.get_month_end()
        manual = datetime.now(tz=pytz.utc)

        self.assertEqual(start.year, manual.year)
        self.assertEqual(start.month, manual.month)
        self.assertEqual(start.day, monthrange(manual.year, manual.month)[1])
        self.assertEqual(start.hour, 23)
        self.assertEqual(start.minute, 59)
        self.assertEqual(start.second, 59)
        self.assertEqual(start.microsecond, 999999)

    def test_get_year_start(self):
        start = TimeUtility.get_year_start()
        manual = datetime.now(tz=pytz.utc)

        self.assertEqual(start.year, manual.year)
        self.assertEqual(start.month, 1)
        self.assertEqual(start.day, 1)
        self.assertEqual(start.hour, 0)
        self.assertEqual(start.minute, 0)
        self.assertEqual(start.second, 0)
        self.assertEqual(start.microsecond, 0)

    def test_get_year_end(self):
        start = TimeUtility.get_year_end()
        manual = datetime.now(tz=pytz.utc)

        self.assertEqual(start.year, manual.year)
        self.assertEqual(start.month, 12)
        self.assertEqual(start.day, 31)
        self.assertEqual(start.hour, 23)
        self.assertEqual(start.minute, 59)
        self.assertEqual(start.second, 59)
        self.assertEqual(start.microsecond, 999999)

    def test_adjust_offset(self):
        utc_time = TimeUtility.now()
        local_time = datetime.now()

        small_time = utc_time
        large_time = local_time

        minute_difference = TimeUtility.difference(large_time, small_time, TimeUtility.MINUTE)

        adjusted = TimeUtility.adjust_offset(utc_time, minute_difference, False)

        aware_local = TimeUtility.make_aware(local_time)

        self.assertEqual(aware_local.year, adjusted.year)
        self.assertEqual(aware_local.month, adjusted.month)
        self.assertEqual(aware_local.day, adjusted.day)
        self.assertEqual(aware_local.hour, adjusted.hour)
        self.assertEqual(aware_local.minute, adjusted.minute)
        self.assertEqual(aware_local.second, adjusted.second)
        
    def test_is_leap_year(self):
        self.assertTrue(TimeUtility.is_leap_year(2020))
        self.assertFalse(TimeUtility.is_leap_year(2019))

    def test_week(self):
        od = date(2021, 8, 13)

        week = TimeUtility.get_week(od)
        self.assertEqual(week.od, od)
        self.assertEqual(week.week_number, 32)
        self.assertEqual(week.week_start, date(2021, 8, 9))
        self.assertEqual(week.week_end, date(2021, 8, 15))

        week_2 = TimeUtility.get_week_by_week_number(2021, 32)
        self.assertEqual(week_2.od, date(2021, 8, 9))
        self.assertEqual(week_2.week_number, 32)
        self.assertEqual(week_2.week_start, date(2021, 8, 9))
        self.assertEqual(week_2.week_end, date(2021, 8, 15))

        week_3 = TimeUtility.get_week_by_week_number(2020, 23)
        self.assertEqual(week_3.od, date(2020, 6, 1))
        self.assertEqual(week_3.week_number, 23)
        self.assertEqual(week_3.week_start, date(2020, 6, 1))
        self.assertEqual(week_3.week_end, date(2020, 6, 7))

    def test_four_week_period(self):
        # 1. The week is neither in the first nor in the last four weeks of the year
        one_date = date(2020, 6, 5)
        one_start_date, one_start_week, one_end_date, one_end_week = TimeUtility.get_four_week_period(one_date)

        self.assertEqual(one_start_week, 21)
        self.assertEqual(one_start_date, date(2020, 5, 18))
        self.assertEqual(one_end_week, 24)
        self.assertEqual(one_end_date, date(2020, 6, 14))

        # 2. First week of the year starts in the previous year, (given date in month of Jan)
        two_date = date(2020, 1, 1)
        two_start_date, two_start_week, two_end_date, two_end_week = TimeUtility.get_four_week_period(two_date)

        self.assertEqual(two_start_week, 1)
        self.assertEqual(two_start_date, date(2020, 1, 1))
        self.assertEqual(two_end_week, 4)
        self.assertEqual(two_end_date, date(2020, 1, 26))

        # 3. First week of the year starts in the previous year, (given date in month of Dec)
        three_date = date(2019, 12, 31)
        three_start_date, three_start_week, three_end_date, three_end_week = TimeUtility.get_four_week_period(three_date)

        self.assertEqual(three_start_week, 49)
        self.assertEqual(three_start_date, date(2019, 12, 2))
        self.assertEqual(three_end_week, 1)
        self.assertEqual(three_end_date, date(2019, 12, 31))

        # 4. First day of the year is the first day of the week 1
        four_date = date(2018, 1, 1)
        four_start_date, four_start_week, four_end_date, four_end_week = TimeUtility.get_four_week_period(four_date)

        self.assertEqual(four_start_week, 1)
        self.assertEqual(four_start_date, date(2018, 1, 1))
        self.assertEqual(four_end_week, 4)
        self.assertEqual(four_end_date, date(2018, 1, 28))

        # 5. The week 1 starts after Jan 1
        five_date = date(2021, 1, 7)
        five_start_date, five_start_week, five_end_date, five_end_week = TimeUtility.get_four_week_period(five_date)

        self.assertEqual(five_start_week, 53)
        self.assertEqual(five_start_date, date(2021, 1, 1))
        self.assertEqual(five_end_week, 4)
        self.assertEqual(five_end_date, date(2021, 1, 31))

        # 6. The first few days of the year is in the last week of the last year
        six_date = date(2021, 1, 1)
        six_start_date, six_start_week, six_end_date, six_end_week = TimeUtility.get_four_week_period(six_date)

        self.assertEqual(six_start_week, 53)
        self.assertEqual(six_start_date, date(2021, 1, 1))
        self.assertEqual(six_end_week, 4)
        self.assertEqual(six_end_date, date(2021, 1, 31))

        # 7. The last week of the year ends before the year
        seven_date = date(2019, 12, 29)
        seven_start_date, seven_start_week, seven_end_date, seven_end_week = TimeUtility.get_four_week_period(seven_date)

        self.assertEqual(seven_start_week, 49)
        self.assertEqual(seven_start_date, date(2019, 12, 2))
        self.assertEqual(seven_end_week, 1)
        self.assertEqual(seven_end_date, date(2019, 12, 31))

        # 8. The last week of the year ends at Dec 31
        eight_date = date(2017, 12, 20)
        eight_start_date, eight_start_week, eight_end_date, eight_end_week = TimeUtility.get_four_week_period(eight_date)

        self.assertEqual(eight_start_week, 49)
        self.assertEqual(eight_start_date, date(2017, 12, 4))
        self.assertEqual(eight_end_week, 52)
        self.assertEqual(eight_end_date, date(2017, 12, 31))

        # 9. The last week of the year ends after Dec 31
        nine_date = date(2020, 12, 31)
        nine_start_date, nine_start_week, nine_end_date, nine_end_week = TimeUtility.get_four_week_period(nine_date)

        self.assertEqual(nine_start_week, 49)
        self.assertEqual(nine_start_date, date(2020, 11, 30))
        self.assertEqual(nine_end_week, 53)
        self.assertEqual(nine_end_date, date(2020, 12, 31))

        # 10. The week number of the given date is 4
        ten_date = date(2021, 1, 27)
        ten_start_date, ten_start_week, ten_end_date, ten_end_week = TimeUtility.get_four_week_period(ten_date)

        self.assertEqual(ten_start_week, 53)
        self.assertEqual(ten_start_date, date(2021, 1, 1))
        self.assertEqual(ten_end_week, 4)
        self.assertEqual(ten_end_date, date(2021, 1, 31))

        # 11. The week number of the given date is the multiple of 4
        eleven_date = date(2021, 2, 27)
        eleven_start_date, eleven_start_week, eleven_end_date, eleven_end_week = TimeUtility.get_four_week_period(eleven_date)

        self.assertEqual(eleven_start_week, 5)
        self.assertEqual(eleven_start_date, date(2021, 2, 1))
        self.assertEqual(eleven_end_week, 8)
        self.assertEqual(eleven_end_date, date(2021, 2, 28))


if __name__ == '__main__':
    unittest.main()


time_utility/week.py
# Python
from datetime import date, timedelta, datetime


class TimeUtilityWeek:

    def __init__(self, od: date):
        self.od: date = od
        self.week_number: int = od.isocalendar()[1]
        self.week_start = od - timedelta(days=od.isoweekday() - 1)
        self.week_end = od + timedelta(days=7 - od.isoweekday())

    @classmethod
    def get_week_from_week_number(cls, year: int, week: int) -> 'TimeUtilityWeek':
        w = "{y}-W{wn}".format(y=year, wn=week)
        d = datetime.strptime(w + '-1', '%G-W%V-%u').date()
        return cls(d)

    @classmethod
    def get_four_week_period(cls, d: date) -> tuple[date, int, date, int]:
        week = cls(d)

        first_week = 1
        last_week = 4

        if week.week_number == 1 and d.month == 12:  # The last few days of year but in the first week of next year
            last_day_of_last_week = d - timedelta(days=1)
            last_week_of_year = cls(last_day_of_last_week)
            first_week = 49
            last_week = 1

            from_d = cls.get_week_from_week_number(d.year, 49).week_start
            to_d = date(week.week_start.year, 12, 31)

        elif week.week_number == 1 and d.month == 1 and week.week_start.month == 12:  # The Week 1 starts before Jan 1 and the selected date is in the new year
            first_week = 1
            last_week = 4
            from_d = date(d.year, 1, 1)
            to_d = cls.get_week_from_week_number(week.week_start.year + 1, 4).week_end

        elif week.week_number == 1 and week.week_start.month == 1 and week.week_start.isoweekday() == 1 and week.week_start.day == 1:  # The first day of year and the first day of the week 1
            first_week = 1
            last_week = 4
            from_d = date(week.week_start.year, 1, 1)
            to_d = date(week.week_start.year, 1, 28)
        elif week.week_number == 1 and week.week_start.month == 1 and week.week_start.isoweekday() > 1:  # The first day of the week 1 starts after Jan 1
            last_day_of_last_week = week.week_start - timedelta(days=1)
            last_week_of_year = cls(last_day_of_last_week)
            first_week = last_week_of_year.week_number
            last_week = 4
            from_d = week.week_start  # date(d.year, 1, 1)
            to_d = cls.get_week_from_week_number(week.week_start.year, 4).week_end
        elif week.week_number >= 52 and d.year > week.week_start.year:  # The first few days of the year in the last week of the last year
            first_week = week.week_number
            last_week = 4
            from_d = date(d.year, 1, 1)
            to_d = cls.get_week_from_week_number(d.year, 4).week_end
        elif week.week_number >= 49:
            first_week = 49
            from_d = cls.get_week_from_week_number(week.week_start.year, 49).week_start
            to_d = date(week.week_start.year, 12, 31)
            last_week = cls(to_d).week_number
        elif week.week_number <= 4:
            from_d = date(week.week_start.year, 1, 1)
            first_week = cls(from_d).week_number
            last_week = 4
            to_d = cls.get_week_from_week_number(week.week_start.year, 4).week_end
        else:
            if week.week_number % 4 == 0:
                first_week = week.week_number - 3
                last_week = week.week_number
            else:
                first_week = week.week_number - ((week.week_number % 4) - 1)
                last_week = week.week_number + (4 - (week.week_number % 4))

            # The first and last calculated weeks
            fw = cls.get_week_from_week_number(d.year, first_week)
            lw = cls.get_week_from_week_number(d.year, last_week)

            from_d = fw.week_start
            to_d = lw.week_end

        return from_d, first_week, to_d, last_week


time_utility/main.py
# Python
from datetime import datetime, timedelta, date
from calendar import monthrange
from typing import Tuple

# Third Party
import pytz

# This Package
from .week import TimeUtilityWeek


class TimeUtility:
    """Date and Time utility functions"""

    # Period Constants
    DAILY = "daily"
    MONTHLY = "monthly"
    ANNUAL = "annual"

    # Time Unit Constants
    MICROSECOND = "microsecond"
    SECOND = "second"
    MINUTE = "minute"
    HOUR = "hour"
    DAY = "day"

    @staticmethod
    def is_naive(target_datetime: datetime) -> bool:
        """
        Checks if the given datetime object is naive or not

        :param target_datetime: The target datetime object to be checked
        """
        return target_datetime.tzinfo is None

    @staticmethod
    def is_aware(target_datetime: datetime) -> bool:
        """
        Checks if the given datetime object is aware or not

        :param target_datetime: The target datetime object to be checked
        """
        return target_datetime.tzinfo is not None

    @staticmethod
    def make_aware(target_datetime: datetime, timezone=pytz.utc) -> datetime:
        """
        Return a new datetime object with the timezone information

        :param target_datetime: The target datetime object
        :param timezone: The target timezone. The default value is UTC
        """
        return target_datetime.replace(tzinfo=timezone)

    @staticmethod
    def now(timezone=pytz.utc) -> datetime:
        """
        Returns the current date and time with the given time zone

        :param timezone: The desired timezone, defaults to UTC
        """
        return datetime.now(tz=timezone)

    @staticmethod
    def today(timezone=pytz.utc) -> date:
        """
        Returns the current date with the given timezone

        :param timezone: The desired timezone, defaults to UTC
        """
        return datetime.now(tz=timezone).date()

    @staticmethod
    def get_date_start(timezone=pytz.utc, year: int | None = None, month: int | None = None, day: int | None = None) -> datetime:
        """
        Returns the date and time of the beginning of the day with the given timezone

        :param timezone: The desired timezone, defaults to UTC
        :param year: The target year, ignore or pass None to use the current year
        :param month: The target month, ignore or pass None to use the current month
        :param day: The target day, ignore or pass None to use the current day
        """
        return datetime.now(tz=timezone).replace(
            year=year if year is not None else datetime.now(tz=timezone).year,
            month=month if month is not None else datetime.now(tz=timezone).month,
            day=day if day is not None else datetime.now(tz=timezone).day,
            hour=0,
            minute=0,
            second=0,
            microsecond=0
        )

    @staticmethod
    def get_date_end(timezone=pytz.utc, year: int | None = None, month: int | None = None, day: int | None = None) -> datetime:
        """
        Returns the date and time of the ending of the day with the given timezone

        :param timezone: The desired timezone, defaults to UTC
        :param year: The target year, ignore or pass None to use the current year
        :param month: The target month, ignore or pass None to use the current month
        :param day: The target day, ignore or pass None to use the current day
        """
        return datetime.now(tz=timezone).replace(
            year=year if year is not None else datetime.now(tz=timezone).year,
            month=month if month is not None else datetime.now(tz=timezone).month,
            day=day if day is not None else datetime.now(tz=timezone).day,
            hour=23,
            minute=59,
            second=59,
            microsecond=999999
        )

    @staticmethod
    def get_month_start(timezone=pytz.utc, year: int | None = None, month: int | None = None) -> datetime:
        """
        Returns the date and time of the beginning of the month with the given timezone

        :param timezone: The desired timezone, defaults to UTC
        :param year: The target year, ignore or pass None to use the current year
        :param month: The target month, ignore or pass None to use the current month
        """
        return datetime.now(tz=timezone).replace(
            year=year if year is not None else datetime.now(tz=timezone).year,
            month=month if month is not None else datetime.now(tz=timezone).month,
            day=1,
            hour=0,
            minute=0,
            second=0,
            microsecond=0
        )

    @staticmethod
    def get_month_end(timezone=pytz.utc, year: int | None = None, month: int | None = None) -> datetime:
        """
        Returns the date and time of the ending of the month with the given timezone

        :param timezone: The desired timezone, defaults to UTC
        :param year: The target year, ignore or pass None to use the current year
        :param month: The target month, ignore or pass None to use the current month
        """
        final_year = year if year is not None else datetime.now(tz=timezone).year
        final_month = month if month is not None else datetime.now(tz=timezone).month
        return datetime.now(tz=timezone).replace(
            year=final_year,
            month=final_month,
            day=monthrange(final_year, final_month)[1],
            hour=23,
            minute=59,
            second=59,
            microsecond=999999
        )

    @staticmethod
    def get_year_start(timezone=pytz.utc, year: int | None = None):
        """
        Returns the date and time of the beginning of the year with the given timezone

        :param timezone: The desired timezone, defaults to UTC
        :param year: The target year, ignore or pass None to use the current year
        """
        return datetime.now(tz=timezone).replace(
            year=year if year is not None else datetime.now(tz=timezone).year,
            month=1,
            day=1,
            hour=0,
            minute=0,
            second=0,
            microsecond=0
        )

    @staticmethod
    def get_year_end(timezone=pytz.utc, year: int | None = None):
        """
        Returns the date and time of the ending of the year with the given timezone

        :param timezone: The desired timezone, defaults to UTC
        :param year: The target year, ignore or pass None to use the current year
        """
        return datetime.now(tz=timezone).replace(
            year=year if year is not None else datetime.now(tz=timezone).year,
            month=12,
            day=31,
            hour=23,
            minute=59,
            second=59,
            microsecond=999999
        )

    @staticmethod
    def adjust_offset(original_datetime: datetime, offset: int, local_to_utc: bool) -> datetime:
        """
        Returns a new datetime object by adjusting the original datetime according to the given offset (in minutes). Note that Javascript offset should be multiplied by -1

        :param original_datetime: The original and unchanged datetime object
        :param offset: The desired offset in minutes (Note that the Javascript offset obtained via `new Date().getTimezoneOffset()` should be multiplied by -1)
        :param local_to_utc: Set to True if the original datetime is in local time and set to False if the original datetime is in UTC timezone
        """
        if local_to_utc:
            new_time = original_datetime - timedelta(minutes=offset)
        else:
            new_time = original_datetime + timedelta(minutes=offset)
        return new_time

    @staticmethod
    def get_period(year: int, month: int, day: int, period: str, offset: int = 0) -> Tuple[datetime, datetime]:
        """
        Returns the beginning and the ending datetime of a period

        :param year: The target year
        :param month: The target month
        :param day: The target day
        :param period: The desired period time-span. Please use the period constants of TimeUtility such as `TimeUtility.DAILY'
        :param offset: The optional timezone offset in minutes. (Note that the Javascript offset obtained via `new Date().getTimezoneOffset()` should be multiplied by -1)
        """
        if period == TimeUtility.DAILY:
            start = TimeUtility.get_date_start(year=year, month=month, day=day)
            end = TimeUtility.get_date_end(year=year, month=month, day=day)
        elif period == TimeUtility.MONTHLY:
            start = TimeUtility.get_month_start(year=year, month=month)
            end = TimeUtility.get_month_end(year=year, month=month)
        elif period == TimeUtility.ANNUAL:
            start = TimeUtility.get_year_start(year=year)
            end = TimeUtility.get_year_end(year=year)
        else:
            raise ValueError()

        if offset == 0:
            return start, end

        return TimeUtility.adjust_offset(start, offset, False), TimeUtility.adjust_offset(end, offset, False)

    @staticmethod
    def difference(large_time: datetime, small_time: datetime, time_span: str = 'second') -> int:
        """
        The difference between two datetime objects in microseconds, seconds, minutes, hours, or days

        :param large_time: The larger datetime object
        :param small_time: The smaller datetime object
        :param time_span: The time-span of difference. Please use the time unit constants of TimeUtility such as `TimeUtility.HOUR`
        """

        final_large = large_time if TimeUtility.is_aware(large_time) else TimeUtility.make_aware(large_time)
        final_small = small_time if TimeUtility.is_aware(small_time) else TimeUtility.make_aware(small_time)

        if time_span == TimeUtility.SECOND:
            return abs((final_large - final_small).seconds)
        elif time_span == TimeUtility.MICROSECOND:
            return abs((final_large - final_small).microseconds)
        elif time_span == TimeUtility.DAY:
            return abs((final_large - final_small).days)
        elif time_span == TimeUtility.MINUTE:
            return int(abs((final_large - final_small).seconds)/60)
        elif time_span == TimeUtility.HOUR:
            return int(abs((final_large - final_small).seconds)/3600)
        else:
            raise ValueError()

    @staticmethod
    def is_leap_year(year: int | None = None):
        """
        Checks whether the given year is a leap year.

        :param year: The target year. Pass None or ignore to use the current year in UTC timezone
        """
        if year is None:
            year = TimeUtility.now().year

        if isinstance(year, int) is False:
            raise ValueError()

        return year % 4 == 0 and (year % 100 != 0 or year % 400 == 0)

    @staticmethod
    def get_week(d: date) -> TimeUtilityWeek:
        return TimeUtilityWeek(d)

    @staticmethod
    def get_week_by_week_number(year: int, week: int) -> TimeUtilityWeek:
        return TimeUtilityWeek.get_week_from_week_number(year, week)

    @staticmethod
    def get_four_week_period(d: date) -> tuple[date, int, date, int]:
        return TimeUtilityWeek.get_four_week_period(d)
